checker/src/tag_domain.rs

```diff
@@ -121,13 +136,14 @@ impl TagDomain {
     pub fn add_tag(&self, tag: Tag, val: BoolDomain) -> Self {
         TagDomain {
             map: self.map.insert(tag, val),
+            default: self.default,
         }
     }
 
     /// Return a lifted Boolean that indicates the presence of `tag` in the tag domain element.
     #[logfn_inputs(TRACE)]
     pub fn has_tag(&self, tag: &Tag) -> BoolDomain {
-        *self.map.get(tag).unwrap_or(&BoolDomain::False)
+        *self.map.get(tag).unwrap_or(&self.default)
```

Comment:
I expect to see BoolDomain::Top here.

Reply:
If a tag is never attached to a value, the tap map will not have an entry for the tag. This is the reason why I put BoolDomain::False previously.

But the change in CallVisitor::handle_has_tag introduces a problem. Previously if a path is not in the environment, we construct an unknown tag check. Now the use of `lookup_path_and_refine_result` may return a variable, whose tag map should indeed be TOP.

Therefore, I want get_tags to return a tag domain element that maps every tag to TOP. To achieve that, I introduce the default field in TagDomain.

Reply:
I think this is too blunt a solution. A given value may have some tags that are known and some that are not yet known.

Reply:
Yes, so the default field is only used to control tags that are not yet known. But there are two different situations:
- A local variable should initially have no tags.
- A function parameter should mark the presence of every tag as "not yet known".

Reply:
Indeed, but the knowledge of how to deal with this obviously lies with the caller of has_tag. You could arguably return Bottom here (which is what default seems to be), but I would prefer Top.

Reply:
Returning either Bottom or Top will presumably fail the following test:
``` rust
let non_secret = 12345;
verify!(does_not_have_tag(&non_secret, Taint));
```

Reply:
Perhaps has_tag should return a bool?

Reply:
It seems that the problem is that we don't known when we don't know.

Reply:
Since tags can never be deleted, the absence of a tag always means no for a local and unknown for a parameter. So the only thing the caller needs to know is "is the tag present?".

Reply:
Well, it might work if has_tag just returns if the tag map contains the entry or not, and all the callers of has_tag do their job to distinguish false and unknown.

Reply:
I guess the issue here is that whether has_tag can already decide the result or not. If has_tag indeed needs information from the callers, we should let the callers make the decision.

Reply:
I suspect that will make things easier to understand. Right now, clearly, things are not particularly easy to understand,

Reply:
I'm not sure whether the caller can easily figure things out. Consider the program below:
``` rust
let temp = local + param;
verify!(has_tag!(&temp, Taint));
```
If the tag sets for both `local` and `param` are empty, the function call to `has_tag` returns "tag is not present".
Then the caller would like to figure out whether the tag is absent or unknown.
But this depends on the propagation behavior of Taint:
- If the addition `+` blocks Taint, then temp does not have Taint.
- Otherwise, if the addition `+` propagates Taint, then whether temp has Taint or not is unknown.